test/rule_definition_tools.jl

```diff
@@ -98,6 +106,25 @@ end
             end
         end
 
+        @testset "Constructors" begin
+            @non_differentiable NonDiffExample(::Any)
+
+            @test isequal(
+                frule((Zero(), 1.2), NonDiffExample, 2.0),
+                (NonDiffExample(2.0), DoesNotExist())
+            )
+
+            res, pullback = rrule(NonDiffExample, 2.0)
+            @test res == NonDiffExample(2.0)
+            @test pullback(1.2) == (NO_FIELDS, DoesNotExist())
+
+            # https://github.com/JuliaDiff/ChainRulesCore.jl/issues/213
+            # problem was that `@nondiff Foo(x)` was also defining rules for other types.
```

Comment:
Wow, that's funky. I guess solves the issue in this comment?
https://github.com/JuliaDiff/ChainRules.jl/pull/310#discussion_r527469358

Reply:
Yep.
Its because `typeof(Foo) == DataType` and `typeof(Bar) == DataType`
but `Core.Typeof(Foo) == Type{Foo}` and `Core.Typeof(Bar) == Type{Bar}`